### ISSUE TITLE
`dlopen` does not make symbols of dependencies visible

### ISSUE THREAD

**Issue by ryanking13** [emscripten-core/emscripten#18264]
**Version of emscripten/emsdk:**

emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.26 (8eaf19f1c6a9a1b0cd0f9a91657366829e34ae5c)
clang version 16.0.0 (https://github.com/llvm/llvm-project f81f0cb75a2808a67d2662f044ad07628fc9d900)

---

`loadDynamicLibrary`, which is internally used in `dlopen`, loads a dynamic library and its dependencies.

When loading dependencies, `loadDynamicLibrary` propagates `flags` parameter as-is. So if a target library is loaded locally (e.g. `loadDynamicLibrary(..., {global: false})` or `dlopen(..., RTLD_NOW | RTLD_LOCAL)`) it also loads its dependencies locally.

~~This is problematic because normally a shared library needs to use symbols from its dependencies.~~

_Edited_

The problem is that currently, a side module cannot see symbols of its dependencies if they are loaded locally. For example, when `a.so` depends on `b.so`, if we load `a.so` locally via `dlopen("a.so", RTLD_LOCAL)`, `a.so` can't use symbols from `b.so`.

### How to reproduce

`main.c` depends on `dep.so` and `dep.so` depends on `dep2.so`. In native (linux), this works fine.

```bash
#!/bin/bash

set -e

cat > dep2.h << EOF
int _add(int a);
EOF

cat > dep2.c << EOF
int _add(int a) { return a; }
EOF

cat > dep.h << EOF
int add(int a);
EOF

cat > dep.c << EOF
#include "dep2.h"
int add(int a) { return _add(a) + 1; }
EOF

cat > main.c << EOF
#include<dlfcn.h>
#include<stdio.h>
#include "dep.h"

int main() {
    void *lib_handle = dlopen("./dep.so", RTLD_NOW | RTLD_LOCAL);
    int (*add)(int) = (int (*)(int))dlsym(lib_handle, "add");

    printf("3 + 1 = %d\n", add(3));
}
EOF

rm -rf dep.o dep2.o dep.so dep2.so main
gcc dep2.c -shared -o dep2.so
gcc dep.c dep2.so -shared -o dep.so
gcc main.c -ldl -o main

LD_LIBRARY_PATH=$(pwd) ./main
```

While in emscripten, the same example raises error,

```bash
#!/bin/bash

set -e

cat > dep2.h << EOF
int _add(int a);
EOF

cat > dep2.c << EOF
int _add(int a) { return a; }
EOF

cat > dep.h << EOF
int add(int a);
EOF

cat > dep.c << EOF
#include "dep2.h"
int add(int a) { return _add(a) + 1; }
EOF

cat > main.c << EOF
#include<dlfcn.h>
#include<stdio.h>
#include "dep.h"

int main() {
    void *lib_handle = dlopen("./dep.so", RTLD_NOW | RTLD_LOCAL);
    int (*add)(int) = (int (*)(int))dlsym(lib_handle, "add");

    printf("3 + 1 = %d\n", add(3));
}
EOF

rm -rf dep.o dep2.o dep.so dep2.so main
emcc -c dep.c
emcc -c dep2.c
emcc dep2.o -shared -o dep2.so -sSIDE_MODULE
emcc dep.o dep2.so -shared -o dep.so -sSIDE_MODULE

emcc main.c -sMAIN_MODULE -o main

node main
```

```
Aborted(Assertion failed: undefined symbol `_add`. perhaps a side module was not linked in?
if this global was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment)
```

**Comment by allsey87** [comment]
@ryanking13 just to confirm my own understanding, the situation *should* be as follows right?

![image](https://github.com/emscripten-core/emscripten/assets/3796345/46d52a71-f36b-40d1-b300-e3cbb67434fb)

**Comment by sbc100** [comment]
@ryanking13  you are correct, in emscripten all symbols from the main module are accessible to all modules that get loaded.   Indeed we ship the whole of libc, libc++ (and all other core libraries) embedded in the main module so pretty much all modules that get loaded end up using some of those symbols.

**Comment by ryanking13** [comment]
@allsey87 Right, and the issue I raised in this post is that if the library is loaded with RTLD_LOCAL, liba.so will not see the symbols in libb.so.

**Comment by sbc100** [comment]
I confirmed that this was fixed in #18924 which was released in 3.1.34
